Staging 20170626

gctucker opened the pull request:
Summary for this update from staging:

* improve lab handling
* add LAVA v2 callback support
* put all Python scripts under LGPL v2.1+
* do some clean up in LAVA v2 scripts
* ~~wait for at least one build to be in KernelCI backend before triggering boots (can happen on slow systems)~~ (should be fixed on the server side instead)
* ~~add LAVA v2 generic qemu test (legacy aarch64 is still there for LAVA v1 jobs)~~ (will be sent again in a separate pull request)

Builds: https://staging.kernelci.org/build/gtucker/branch/kernelci-test/kernel/kernelci-test-001-10-g86ca3542cc52/
Boots: https://lava.collabora.co.uk/scheduler/alljobs?search=kernelci-test-001-10

Note: Please try not to squash all these commits into one when merging the pull request as doing this makes it much harder to work with the history.

gctucker committed:
lava-v2-jobs-from-api.py: reformat job dict with one item per line

Reformat the 'job' dictionary with only one item per line for
readability, with no functional change.

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

gctucker committed:
lava-v2-jobs-from-api.py: add git_url to the meta-data

Add the git_url to the jobs meta-data so that the KernelCI front-end
can provide links to the kernel binaries.

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

gctucker committed:
lava-v2-jobs-from-api.py: refactor INITRD_URL and NFSROOTFS_URL

Instead of duplicating all the URLs, define a list of ARCHS and
strings where the arch gets replaced.

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

gctucker committed:
lava-v2-jobs-from-api.py: use urllib.urlencode to encode URL params

Use the urllib.urlencode method to "percent-encode" the API URL and
avoid potential issues with special characters in particular.

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

gctucker committed:
lava-v2-jobs-from-api.py: use response.raise_for_status

Use the standard response.raise_for_status method to raise an
exception if there is an HTTP error in the response.

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

gctucker committed:
lava-v2-submit-jobs.py: use yaml.safe_load instead of yaml.load

Use yaml.safe_load to avoid potential security issues, as documented:

  http://pyyaml.org/wiki/PyYAMLDocumentation#LoadingYAML

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

Commit:
Make all LAVA_LABNAME_TOKEN variables common

- enable --callback for all labs

Commit:
Make free-electrons v2 jobs be medium priority

Commit:
Handle staging lab names (like lab-mhart-dev)

Commit:
Support changing the callback token name

gctucker committed:
Add LGPL v2.1+ headers in Python scripts and full license in COPYING

Update all the Python scripts with a header listing the shared
copyrights, authors and a notice to distribute the code under LGPL
v2.1 or later.  Also add the full license text in COPYING.

Note: Some scripts were already in GPLv2 so they remain untouched.
Also, log2html.py is a temporary local copy - see lava-report.py.

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

gctucker commented:
Noticed an issue with this boot (qemu x86):
https://lava.collabora.co.uk/scheduler/job/755869

mattface commented on lava-job-runner.py:
```diff
@@ -1,4 +1,22 @@
 #!/usr/bin/python
+#
+# Copyright (C) 2014, 2015, 2016 Linaro Limited
```
I'd quite like to have actual authors here, maybe leave the licensing stuff out of this until the rest of the team are back from wherever they've been

gctucker replied:
OK, it's not hard to add a list of authors as the history is not very long.  I'd rather add it and update this patch than start chasing everyone until they've added a couple of lines each...

mattface replied:
I'm not one of the founders of kci so I'd want Tyler, Milo or Kevin to approve the license changes before I'd merge this

gctucker replied:
Sure, so I'll push an updated change with all the authors, then ask the founders to confirm they're happy with the license (should be fine as it has been previously discussed).

mattface commented on lava-v2-jobs-from-api.py:
```diff
@@ -15,22 +33,16 @@
-
+ARCHS = ['arm64', 'arm64be', 'armeb', 'armel', 'x86']
+ROOTFS_URL = 'http://storage.kernelci.org/images/rootfs'
+INITRD_URL = '/'.join([ROOTFS_URL, 'buildroot/{}/rootfs.cpio.gz'])
```
Feels like this is removing flexibility and it was fine as it was

gctucker replied:
I found the original list quite hard to manage, for example if you want to use binaries from a different place you then have to edit each line.  It's easy to get a typo wrong in a list like that.

If it's only me then I can drop this change, let me know.

mattface replied:
Not worth the effort to redo it, leave your change in

mattface commented on lava-v2-jobs-from-api.py:
```diff
@@ -58,17 +70,33 @@ def main(args):
+    poll_seconds = 10
+    loops = 600 / poll_seconds # time out after 10 minutes
+    loop = 0
+    while not builds and loop < loops:
```
Polling? Why? This is only executed by jenkins once all builds for an arch have finished. If there are no builds something has gone wrong with the jenkins setup and this should exit immediately.

gctucker replied:
This was discussed on IRC, it's a real issue.  When a boot report has been sent to the KernelCI backend, an asynchronous task is being run (with Celery).  So the boot data is not in the database at the end of the request, it only gets added later on.  It is usually fast but there's nothing to guarantee when the boot data will actually be in the database.

I could hit this issue regularly on my set-up running in a VM, it could sometimes take a few seconds between the end of the "boot POST" request and when the data is actually available in the database.  Calling the lava-v2-jobs-from-api.py script too early can cause some kernel builds to not be boot tested.

mattface replied:
Build data right?

If it's possible that the builds arent all in the database by the time this script comes to use it, we need to fix that rather than polling, because how do you know when polling that you've got all the builds? This code would determine 1 result is enough and create jobs for it.

We do many builds per arch so the first one will be in the db well before the last one is finished, maybe we need to know how the longest time the celery cache can be out of date and just wait that long?

gctucker replied:
I've discussed this with Milo, I'll drop this polling patch for now and try to fix it on the server side by making the API synchronous.  Doing this should solve the problem, but we need to check it doesn't cause any bad side-effects or any impact on performance.

It's true that polling for only one job is not completely correct either, I guess I didn't really want to expand too much on this work-around.  Hopefully the synchronous API approach will just work.

mattface commented on lava-v2-jobs-from-api.py:
```diff
@@ -96,10 +124,9 @@ def main(args):
                         exit(1)
             if build['kernel_image']:
                 # handle devices without a DTB, hacky :/
+                build['dtb_dir_data'].append('qemu')
```
v1 doesn't run a qemu job for every single defconfig, but thats what this would change it to, excluding the blacklist in the device_map.

v1 uses the vexpress-v2p-ca{9/15} dtbs in the device_map to know if a defconfig should be booted by qemu, and the qemu arm64 was only booted on defconfig*

Also looks like x86 qemu boots need the arch in the job file to be x86_64 to work properly

gctucker replied:
Yes the link I put in my previous comment is about the arch in the job context, I've got a fix for that.

With LAVA v1, a different device type was needed for each qemu arch.  This is not the case anymore, so I think things can be consolidated a bit better for v2.  It's been useful for me to run on qemu to test the changes I'm working on, I'll look a bit more into this and send an updated patch.

mattface replied:
Yes to consolidation, just having a qemu device for v2 is nice

Maybe we actually do want a qemu boot on every defconfig but we haven't previously so this is a change of behaviour. I've actually turned off v2 boots on production as there's no point, but this would produce a lot more boot jobs for all labs when it does get turned on so it's something to discuss.

gctucker replied:
Sorry about the defconfig bit, that's because I was testing on a subset of the whole KernelCI coverage (which is too big to keep testing in practice).  I'll drop this patch from this pull request and submit a cleaner one later on.

PS: Although it's not used in production yet, I am using only v2 boots with my dev system :)

khilman replied:
Re: QEMU boots in every lab.  Personally, I don't want my lab bogged down with QEMU boots, so I won't setup QEMU devices.

I think we should use lava-docker to setup a lab in the cloud that does all the QEMU boots so that real labs can focus on real hardware.

gctucker replied:
We have a couple of qemu devices on a separate dispatcher in the Collabora lab, so it's fine for us to run some jobs.  There aren't that many to run in fact with the limited defconfig set, and they're quick.  Having a cloud lab would still help I guess, to provide some redundancy.

I suppose you could still have qemu devices in you LAVA lab but not make them available to kernelci.  Or maybe this is another case of per-lab filtering (trees and device types).

mattface commented on lava-v2-jobs-from-api.py:
```diff
@@ -157,32 +184,59 @@ def main(args):
-                                        else:
-                                            nfsrootfs_url = None
+                                        initrd_url = INITRD_URL.format(initrd_arch)
+                                        nfsrootfs_url = NFSROOTFS_URL.format(initrd_arch) if 'nfs' in plan else None
```
I personally don't like this sort of layout, especially as there are plenty of IF ELSE blocks elsewhere that don't look like this.
Can leave this change in for now, but I'm not really interested in cleanup commits at this stage.

gctucker replied:
Well, I personally find it better to have one line to define the value of a variable.  Can revert that if it's breaking the coding style, it's part of the change to refactor the list of URLs.

mattface replied:
As it's used with the url changes, leave it in as it's not worth the effort to remove now

gctucker replied:
Fine.

mattface reviewed (commented):
The qemu stuff seems like it would run a lot more jobs than v1 did, and the license stuff needs more input, especially as I hadn't merged the LGPL pull request into lava-ci-staging.

gctucker commented:
This is based on the `staging-20170627` tag from lava-ci-staging.  It addresses all the comments above.

Builds: https://staging.kernelci.org/build/gtucker/branch/kernelci-test/kernel/kernelci-test-001-17-g3ec98f5b1c1f/
Boots: https://lava.collabora.co.uk/scheduler/alljobs?search=kernelci-test-001-17
